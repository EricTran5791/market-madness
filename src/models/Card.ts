import { types } from 'mobx-state-tree';
import {
  CardEffectUnion,
  CardEffectKind,
  BasicCardEffectSnapshotType,
  CardEffectCategory,
  InteractiveCardEffectSnapshotType,
  InteractiveCardEffectCategory,
} from './CardEffect';

export enum CardCategory {
  Action = 'Action',
  Consumable = 'Consumable',
  Item = 'Item',
  Money = 'Money',
  NPC = 'NPC',
}

export enum CardSubcategory {
  Food = 'Food',
}

export const Card = types
  .model({
    id: types.identifier(types.string),
    name: types.string,
    category: types.enumeration(
      'Category',
      Object.keys(CardCategory).map(key => CardCategory[key])
    ),
    subcategories: types.optional(
      types.array(
        types.enumeration(
          'Subcategory',
          Object.keys(CardSubcategory).map(key => CardSubcategory[key])
        )
      ),
      []
    ),
    description: types.optional(types.string, ''),
    cost: types.optional(types.number, 0),
    money: types.optional(types.number, 0),
    health: types.optional(types.number, 0),
    isPlayed: types.optional(types.boolean, false),
    effects: types.optional(types.array(CardEffectUnion), []),
  })
  .actions(self => ({
    afterCreate() {
      if (self.description.length > 0 || self.effects.length === 0) {
        return;
      }

      // Auto generate descriptions
      if (self.category === CardCategory.NPC) {
        self.description = 'Defeat: ';
      }
      self.description += self.effects
        .map(effect => {
          if (effect.kind === CardEffectKind.Basic) {
            const {
              category,
              value,
              gainedCardId,
            }: BasicCardEffectSnapshotType = effect;
            switch (category) {
              case CardEffectCategory.Draw:
                return `Draw ${value} card${value > 1 ? 's' : ''}`;
              case CardEffectCategory.GainAttack:
                return `+${value} Attack`;
              case CardEffectCategory.GainCardToDiscardPile:
                // The RegEx converts the camel case id to a spaced and capitalized name
                // TODO: There was a circular dependency when cardGenerator was imported
                // due to cardLibrary also importing CardCategory from Card
                return `Gain ${value} ${gainedCardId
                  .replace(/([A-Z])/g, ' $1')
                  .replace(/^./, (str: string) => {
                    return str.toUpperCase();
                  })}`;
              case CardEffectCategory.GainCardToHand:
                // The RegEx converts the camel case id to a spaced and capitalized name
                // TODO: There was a circular dependency when cardGenerator was imported
                // due to cardLibrary also importing CardCategory from Card
                return `Add ${value} ${gainedCardId
                  .replace(/([A-Z])/g, ' $1')
                  .replace(/^./, (str: string) => {
                    return str.toUpperCase();
                  })} to your hand`;
              case CardEffectCategory.GainMoney:
                return `+${value} Money`;
              case CardEffectCategory.Heal:
                return `Heal ${value}`;
              case CardEffectCategory.IncreaseMaxHealth:
                return `Increase max health by ${value}`;
              case CardEffectCategory.TrashSelf:
                return `Trash this card`;
              default:
                return '';
            }
          }
          if (effect.kind === CardEffectKind.Interactive) {
            const {
              category,
              numCardsToResolve,
            }: InteractiveCardEffectSnapshotType = effect;
            switch (category) {
              case InteractiveCardEffectCategory.OptionalTrash:
                return `Trash up to ${numCardsToResolve} cards`;
              case InteractiveCardEffectCategory.MandatoryDiscard:
                return `Discard ${numCardsToResolve} cards`;
              case InteractiveCardEffectCategory.MandatoryTrash:
                return `Trash ${numCardsToResolve} cards`;
              default:
                return '';
            }
          }
          return '';
        })
        .join(', ');
    },
    /** Reset a card's isPlayed status every time it is detached from its direct parent. */
    beforeDetach() {
      self.isPlayed = false;
    },
  }));

export type CardModelType = typeof Card.Type;
export type CardModelSnapshotType = typeof Card.SnapshotType;

export const CardStack = types
  .model({
    cards: types.array(Card),
  })
  .views(self => ({
    get totalCards(): number {
      return self.cards.length;
    },
    get totalMoney() {
      return self.cards
        .map(card => card.money)
        .reduce((sum, currentValue) => sum + currentValue, 0);
    },
    get unplayedCards() {
      return self.cards.filter(card => !card.isPlayed);
    },
  }))
  .actions(self => ({
    add(card: CardModelType) {
      self.cards.push(card);
    },
    remove(card: CardModelType) {
      self.cards.remove(card);
    },
  }));

export type CardStackModelType = typeof CardStack.Type;
