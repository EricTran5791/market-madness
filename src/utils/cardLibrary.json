{
  "shoppersFrenzy": {
    "id": "shoppersFrenzy",
    "name": "Shopper's Frenzy",
    "kind": "Normal",
    "shop": "Neutral",
    "subcategories": [],
    "description": "+1 Attack, Draw 1 card",
    "cost": {
      "kind": "Money",
      "value": 2
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Gain Attack",
        "value": 1,
        "gainedCard": null
      },
      {
        "kind": "Basic",
        "category": "Draw",
        "value": 1,
        "gainedCard": null
      }
    ]
  },
  "basketball": {
    "id": "basketball",
    "name": "Basketball",
    "kind": "Normal",
    "shop": "Sports",
    "subcategories": ["Sports Equipment"],
    "description": "+3 Attack",
    "cost": {
      "kind": "Money",
      "value": 3
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Gain Attack",
        "value": 3,
        "gainedCard": null
      }
    ]
  },
  "postalWorker": {
    "id": "postalWorker",
    "name": "Postal Worker",
    "kind": "NPC",
    "shop": "Neutral",
    "subcategories": [],
    "description": "Defeat: Draw 1 card",
    "cost": {
      "kind": "Health",
      "value": 2
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Draw",
        "value": 1,
        "gainedCard": null
      }
    ]
  },
  "shoppingBasket": {
    "id": "shoppingBasket",
    "name": "Shopping Basket",
    "kind": "Normal",
    "shop": "Neutral",
    "subcategories": [],
    "description": "+2 Attack",
    "cost": {
      "kind": "Money",
      "value": 2
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Gain Attack",
        "value": 2,
        "gainedCard": null
      }
    ]
  },
  "gem": {
    "id": "gem",
    "name": "Gem",
    "kind": "Money",
    "shop": "Neutral",
    "subcategories": [],
    "description": "+2 Money",
    "cost": {
      "kind": "Money",
      "value": 3
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Gain Money",
        "value": 2,
        "gainedCard": null
      }
    ]
  },
  "coin": {
    "id": "coin",
    "name": "Coin",
    "kind": "Money",
    "shop": "Neutral",
    "subcategories": [],
    "description": "+1 Money",
    "cost": {
      "kind": "Money",
      "value": 1
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Gain Money",
        "value": 1,
        "gainedCard": null
      }
    ]
  },
  "exchangeGoods": {
    "id": "exchangeGoods",
    "name": "Exchange Goods",
    "kind": "Normal",
    "shop": "Neutral",
    "subcategories": [],
    "description": "Draw 2 cards, Discard 2 cards",
    "cost": {
      "kind": "Money",
      "value": 2
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Draw",
        "value": 2,
        "gainedCard": null
      },
      {
        "kind": "Interactive",
        "category": "Discard",
        "numCardsToResolve": 2,
        "resolveType": "Mandatory"
      }
    ]
  },
  "garbageBag": {
    "id": "garbageBag",
    "name": "Garbage Bag",
    "kind": "Normal",
    "shop": "Neutral",
    "subcategories": [],
    "description": "Trash up to 2 cards, Trash this card",
    "cost": {
      "kind": "Money",
      "value": 2
    },
    "effects": [
      {
        "kind": "Interactive",
        "category": "Trash",
        "numCardsToResolve": 2,
        "resolveType": "Optional"
      },
      {
        "kind": "Basic",
        "category": "Trash Self",
        "value": 0,
        "gainedCard": null
      }
    ]
  },
  "businessPerson": {
    "id": "businessPerson",
    "name": "Business Person",
    "kind": "NPC",
    "shop": "Neutral",
    "subcategories": [],
    "description": "Defeat: +3 Money",
    "cost": {
      "kind": "Health",
      "value": 4
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Gain Money",
        "value": 3,
        "gainedCard": null
      }
    ]
  },
  "fishBones": {
    "id": "fishBones",
    "name": "Fish Bones",
    "kind": "Normal",
    "shop": "Neutral",
    "subcategories": [],
    "description": "",
    "cost": {
      "kind": "Money",
      "value": 0
    },
    "effects": []
  },
  "bread": {
    "id": "bread",
    "name": "Bread",
    "kind": "Normal",
    "shop": "Groceries",
    "subcategories": ["Food"],
    "description": "Heal 2",
    "cost": {
      "kind": "Money",
      "value": 2
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Heal",
        "value": 2,
        "gainedCard": null
      }
    ]
  },
  "baker": {
    "id": "baker",
    "name": "Baker",
    "kind": "NPC",
    "shop": "Neutral",
    "subcategories": [],
    "description": "Defeat: Add 1 Bread to your hand",
    "cost": {
      "kind": "Health",
      "value": 3
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Gain Card to Hand",
        "value": 1,
        "gainedCard": {
          "id": "bread",
          "name": "Bread"
        }
      }
    ]
  },
  "portableFurnace": {
    "id": "portableFurnace",
    "name": "Portable Furnace",
    "kind": "Normal",
    "shop": "Artisanal",
    "subcategories": [],
    "description": "Trash 2 cards, Add 1 Gem to your hand",
    "cost": {
      "kind": "Money",
      "value": 4
    },
    "effects": [
      {
        "kind": "Interactive",
        "category": "Trash",
        "numCardsToResolve": 2,
        "resolveType": "Mandatory"
      },
      {
        "kind": "Basic",
        "category": "Gain Card to Hand",
        "value": 1,
        "gainedCard": {
          "id": "gem",
          "name": "Gem"
        }
      }
    ]
  },
  "expressShipping": {
    "id": "expressShipping",
    "name": "Express Shipping",
    "kind": "Normal",
    "shop": "Neutral",
    "subcategories": [],
    "description": "Draw 2 cards",
    "cost": {
      "kind": "Money",
      "value": 3
    },
    "effects": [
      {
        "kind": "Basic",
        "category": "Draw",
        "value": 2,
        "gainedCard": null
      }
    ]
  }
}
