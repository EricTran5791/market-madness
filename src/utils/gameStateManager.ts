import { StoreType, Store } from '../models/Store';
import { Player, PlayerId } from '../models/Player';
import { GameState, GamePhase } from '../models/GameState';
import { onAction, onPatch, getSnapshot, applySnapshot } from 'mobx-state-tree';
import {
  generateMarketDeck,
  generateStartingDeck,
  generateEmptyDeck,
  shuffleCardStackModel,
} from './cardGenerator';

export function initializeStore(): StoreType {
  const players = [
    Player.create({
      id: PlayerId.Player1,
      discardPile: generateEmptyDeck(),
      hand: {
        cardStack: generateEmptyDeck(),
        gainedCardStack: generateEmptyDeck(),
      },
      deck: generateStartingDeck(),
    }),
    Player.create({
      id: PlayerId.Computer,
      discardPile: generateEmptyDeck(),
      hand: {
        cardStack: generateEmptyDeck(),
        gainedCardStack: generateEmptyDeck(),
      },
      deck: generateStartingDeck(),
    }),
  ];

  const gameState = GameState.create({
    currentGamePhase: GamePhase.PlayersTurn,
    currentPlayer: players[0],
  });

  const store = Store.create({
    market: { cardStack: generateMarketDeck() },
    players,
    gameState,
  });

  setupReactions(store);
  store.createNewGame();
  return store;
}

function setupReactions(store: StoreType) {
  const initialSnapshot = getSnapshot(store);

  onAction(store, call => {
    if (call.name === 'createNewGame') {
      applySnapshot(store, initialSnapshot);
      shuffleCardStackModel(store.market.cardStack);
      shuffleCardStackModel(store.currentPlayer.deck);
      shuffleCardStackModel(store.otherPlayer.deck);
      store.currentPlayer.drawFromDeck(5);
    }

    if (call.name === 'endTurn') {
      store.currentPlayer.clearHand();
      store.changeCurrentPlayer();
      store.currentPlayer.drawFromDeck(5);
    }
  });

  // End the game whenever either player reaches 0 health or below
  onPatch(store.players[0], call => {
    if (call.op === 'replace' && call.path === '/health' && call.value <= 0) {
      console.log('game over');
      store.gameState.currentGamePhase = GamePhase.GameOver;
    }
  });
  onPatch(store.players[1], call => {
    if (call.op === 'replace' && call.path === '/health' && call.value <= 0) {
      console.log('game over');
      store.gameState.currentGamePhase = GamePhase.GameOver;
    }
  });
}
