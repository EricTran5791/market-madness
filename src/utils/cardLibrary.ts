import { CardCategory } from '../models/Card';

export const ActionCards = {
  apple: {
    name: 'Apple',
    category: CardCategory.action,
    cost: 1,
  },
  bananas: {
    name: 'Bananas',
    category: CardCategory.action,
    cost: 2,
  },
  couponBook: {
    name: 'Coupon Book',
    category: CardCategory.action,
    cost: 3,
  },
  dragonFruit: {
    name: 'Dragon Fruit',
    category: CardCategory.action,
    cost: 4,
  },
};

export const AttackCards = {
  punch: {
    name: 'Punch',
    category: CardCategory.attack,
    cost: 1,
    attackValue: 1,
  },
};

export const MoneyCards = {
  coin: {
    name: 'Coin',
    category: CardCategory.money,
    cost: 0,
    buyingPower: 1,
  },
  gem: {
    name: 'Gem',
    category: CardCategory.money,
    cost: 0,
    buyingPower: 2,
  },
};
