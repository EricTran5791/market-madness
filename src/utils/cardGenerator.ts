import {
  CardStack,
  CardStackModelType,
  Card,
  CardModelType,
} from '../models/Card';
import * as cards from './cardLibrary';
import * as uniqid from 'uniqid';
import {
  CardEffect,
  CardEffectKind,
  CardEffectCategory,
  InteractiveCardEffectCategory,
  InteractiveCardEffectResolveType,
} from '../types/cardEffect.types';
import { List } from 'immutable';
import { Card as CardType, CardCategory } from '../types/cardTypes';

export function generateEmptyDeck(): CardStackModelType {
  return CardStack.create({ cards: [] });
}

export function printCard(card: CardType): CardModelType {
  return Card.create({
    id: uniqid(),
    ...card,
  });
}

// TODO: Improve with JSON card database
export function printCardById(id: string): CardModelType {
  const card =
    cards.ActionCards[id] ||
    cards.ConsumableCards[id] ||
    cards.ItemCards[id] ||
    cards.MoneyCards[id];
  return printCard(card);
}

function printDuplicateCards(card: CardType, qty: number): CardModelType[] {
  return new Array(qty).fill(undefined).map(_ => printCard(card));
}

export function generateStartingDeck(): CardStackModelType {
  return CardStack.create({
    cards: shuffle([
      ...printDuplicateCards(cards.MoneyCards.coin, 8),
      ...printDuplicateCards(cards.ActionCards.dropShipment, 2),
    ]),
  });
}

export function generateTrashDeck(): CardStackModelType {
  return CardStack.create({
    cards: [printCard(cards.ItemCards.fishBones)],
  });
}

export function generateMarketDeck(): CardStackModelType {
  return CardStack.create({
    cards: shuffle([
      ...printDuplicateCards(cards.ActionCards.exchangeGoods, 3),
      ...printDuplicateCards(cards.ActionCards.expressShipping, 2),
      ...printDuplicateCards(cards.ConsumableCards.energyDrink, 2),
      ...printDuplicateCards(cards.ItemCards.basketball, 3),
      ...printDuplicateCards(cards.ItemCards.garbageBag, 2),
      ...printDuplicateCards(cards.ItemCards.portableFurnace, 2),
      ...printDuplicateCards(cards.NPCCards.baker, 1),
      ...printDuplicateCards(cards.NPCCards.businessPerson, 1),
      ...printDuplicateCards(cards.NPCCards.dietician, 1),
      ...printDuplicateCards(cards.NPCCards.janitor, 1),
      ...printDuplicateCards(cards.NPCCards.postalWorker, 2),
    ]),
  });
}

export function generateAlwaysAvailableDeck(): CardStackModelType {
  return CardStack.create({
    cards: [
      printCard(cards.MoneyCards.gem),
      printCard(cards.ItemCards.shoppingBasket),
    ],
  });
}

// Durstenfeld Shuffle
function shuffle<T>(array: T[]): T[] {
  for (let i = array.length - 1; i > 0; i--) {
    let j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}

export function shuffleCardStackModel(
  cardStack: CardStackModelType
): CardStackModelType {
  for (let i = cardStack.cards.length - 1; i > 0; i--) {
    let j = Math.floor(Math.random() * (i + 1));
    cardStack.cards.move(i, j);
  }
  return cardStack;
}

export function generateCardDescription(
  cardCategory: CardCategory,
  effects: List<CardEffect>
): string {
  return (
    `${cardCategory === CardCategory.NPC ? 'Defeat: ' : ''}` +
    effects
      .map((effect: CardEffect) => {
        if (effect.kind === CardEffectKind.Basic) {
          const { category, value, gainedCardId } = effect;
          switch (category) {
            case CardEffectCategory.Draw:
              return `Draw ${value} card${value > 1 ? 's' : ''}`;
            case CardEffectCategory.GainAttack:
              return `+${value} Attack`;
            case CardEffectCategory.GainCardToDiscardPile:
              // TODO: Map card id to card name once the JSON card library is done
              return `Gain ${value} ${gainedCardId}`;
            case CardEffectCategory.GainCardToHand:
              // TODO: Map card id to card name once the JSON card library is done
              return `Add ${value} ${gainedCardId} to your hand`;
            case CardEffectCategory.GainMoney:
              return `+${value} Money`;
            case CardEffectCategory.Heal:
              return `Heal ${value}`;
            case CardEffectCategory.IncreaseMaxHealth:
              return `Increase max health by ${value}`;
            case CardEffectCategory.TrashSelf:
              return `Trash this card`;
            default:
              console.error(
                `Error: No description template available for basic card effect category '${category}'.`
              );
              return '';
          }
        } else if (effect.kind === CardEffectKind.Interactive) {
          const { category, numCardsToResolve, resolveType } = effect;
          const optionalText =
            resolveType === InteractiveCardEffectResolveType.Optional
              ? 'up to '
              : '';
          switch (category) {
            case InteractiveCardEffectCategory.Discard:
              return `Discard ${optionalText}${numCardsToResolve} cards`;
            case InteractiveCardEffectCategory.Trash:
              return `Trash ${optionalText}${numCardsToResolve} cards`;
            default:
              console.error(
                `Error: No description template available for interactive card effect category '${category}'.`
              );
              return '';
          }
        }
        return '';
      })
      .join(', ')
  );
}
