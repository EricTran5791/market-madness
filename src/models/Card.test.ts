import { Card, CardModelType, CardStack, CardStackModelType } from './Card';
import * as cards from '../utils/cardLibrary';
import { getSnapshot } from 'mobx-state-tree';

let mockCard: CardModelType;
let mockCardStack: CardStackModelType;

beforeEach(() => {
  mockCard = Card.create(cards.ActionCards.apple);
  mockCardStack = CardStack.create({
    cards: [mockCard],
  });
});

it('creates a Card', () => {
  expect(getSnapshot(mockCard)).toMatchSnapshot();
});

it('creates a CardStack', () => {
  expect(getSnapshot(mockCardStack)).toMatchSnapshot();
});

it('adds a card to the CardStack', () => {
  mockCardStack.add(Card.create(cards.ActionCards.bananas));
  expect(getSnapshot(mockCardStack)).toMatchSnapshot();
});

it('removes a card from the CardStack', () => {
  mockCardStack.remove(mockCard);
  expect(getSnapshot(mockCardStack)).toMatchSnapshot();
});
