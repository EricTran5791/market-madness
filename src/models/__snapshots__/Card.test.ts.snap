// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds a card to the CardStack 1`] = `
Object {
  "cards": Array [
    Object {
      "buyingPower": 0,
      "category": "Consumable",
      "cost": 1,
      "description": "Heal 1, Trash this card",
      "effects": Array [
        Object {
          "category": "Heal",
          "kind": "Basic",
          "value": 1,
        },
        Object {
          "category": "Trash Self",
          "kind": "Basic",
          "value": 0,
        },
      ],
      "id": "mock-id",
      "isPlayed": false,
      "name": "Apple",
    },
    Object {
      "buyingPower": 0,
      "category": "Consumable",
      "cost": 2,
      "description": "Heal 2, Trash this card",
      "effects": Array [
        Object {
          "category": "Heal",
          "kind": "Basic",
          "value": 2,
        },
        Object {
          "category": "Trash Self",
          "kind": "Basic",
          "value": 0,
        },
      ],
      "id": "mock-id-2",
      "isPlayed": false,
      "name": "Bananas",
    },
  ],
}
`;

exports[`creates a Card 1`] = `
Object {
  "buyingPower": 0,
  "category": "Consumable",
  "cost": 1,
  "description": "Heal 1, Trash this card",
  "effects": Array [
    Object {
      "category": "Heal",
      "kind": "Basic",
      "value": 1,
    },
    Object {
      "category": "Trash Self",
      "kind": "Basic",
      "value": 0,
    },
  ],
  "id": "mock-id",
  "isPlayed": false,
  "name": "Apple",
}
`;

exports[`creates a CardStack 1`] = `
Object {
  "cards": Array [
    Object {
      "buyingPower": 0,
      "category": "Consumable",
      "cost": 1,
      "description": "Heal 1, Trash this card",
      "effects": Array [
        Object {
          "category": "Heal",
          "kind": "Basic",
          "value": 1,
        },
        Object {
          "category": "Trash Self",
          "kind": "Basic",
          "value": 0,
        },
      ],
      "id": "mock-id",
      "isPlayed": false,
      "name": "Apple",
    },
  ],
}
`;

exports[`removes a card from the CardStack 1`] = `
Object {
  "cards": Array [],
}
`;
